build.zig

```diff
@@ -353,8 +353,63 @@ pub fn build(b: *std.Build) void {
 
     {
         const jni_tests_step = b.step("test:jni", "Run the JNI tests");
-        const jni_tests = JniTestStep.add(b, mode, target);
-        jni_tests_step.dependOn(&jni_tests.step);
+
+        // We need libjvm.so both at build time and at a runtime, so use `FailStep` when that is not
+        // available.
+        if (b.env_map.get("JAVA_HOME")) |java_home| {
+            const libjvm_path = b.pathJoin(&.{
+                java_home,
+                if (builtin.os.tag == .windows) "/lib" else "/lib/server",
+            });
+
+            const tests = b.addTest(.{
+                .root_source_file = .{ .path = "src/clients/java/src/jni_tests.zig" },
+                .target = target,
+                // TODO(zig): The function `JNI_CreateJavaVM` tries to detect
+                // the stack size and causes a SEGV that is handled by Zig's panic handler.
+                // https://bugzilla.redhat.com/show_bug.cgi?id=1572811#c7
+                //
+                // The workaround is run the tests in "ReleaseFast" mode.
+                .optimize = if (builtin.os.tag == .windows) .ReleaseFast else mode,
+            });
+            tests.linkLibC();
+
+            tests.linkSystemLibrary("jvm");
+            tests.addLibraryPath(.{ .path = libjvm_path });
+            if (builtin.os.tag == .linux) {
+                // On Linux, detects the abi by calling `ldd` to check if
+                // the libjvm.so is linked against libc or musl.
+                // It's reasonable to assume that ldd will be present.
+                const ldd_result = b.exec(&.{
+                    "ldd",
+                    b.pathJoin(&.{ libjvm_path, "libjvm.so" }),
+                });
+                tests.target.abi = if (std.mem.indexOf(u8, ldd_result, "musl") != null)
+                    .musl
+                else if (std.mem.indexOf(u8, ldd_result, "libc") != null)
+                    .gnu
+                else {
+                    std.log.err("{s}", .{ldd_result});
+                    return error.JavaAbiUnrecognized;
+                };
+            }
+
+            switch (builtin.os.tag) {
+                .windows => set_windows_dll(b.allocator, java_home),
+                .macos => try b.env_map.put("DYLD_LIBRARY_PATH", libjvm_path),
+                .linux => try b.env_map.put("LD_LIBRARY_PATH", libjvm_path),
```

Comment:
NB: `LD_LIBRARY_PATH` on linux is new, but I think we need that, as JAVA_HOME in general might not be on the runtime search path.